### PR TITLE
Add Rust Oak client library

### PR DESCRIPTION
**Pull request by ipetr0v**
This change adds a Rust `oak_client` library.

Ref https://github.com/project-oak/oak/issues/1097

### CODE REVIEW

File: oak_abi/src/label/mod.rs [hunk 1]
```diff
@@ -54,6 +54,25 @@ impl crate::proto::oak::label::Label {
...
+    /// Returns a WebAssembly module Ed25519 public key label.
+    /// https://ed25519.cr.yp.to
+    pub fn public_key(public_key: &[u8]) -> Self {
+        let public_key_tag = web_assembly_module_signature_tag(&public_key);
+        Label {
+            confidentiality_tags: vec![public_key_tag],
+            integrity_tags: vec![],
+        }
+    }
+
```
Comment (tiziano88) [inline oak_abi/src/label/mod.rs:75]: I don't think these methods should be here (or exist at all). AFAICT they are just wrapping a tag into a label by using it as confidentiality component only. How about a `confidentiality_label` method (outside of `Label`, as top-level, like the existing `web_assembly_module_signature_tag` & similar methods). Also the naming is quite vague, and inconsistent with the existing methods about tags, or is there a reason for that?

File: oak_client/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Intercepts gRPC messages and adds `label` as a gRPC metadata.
+pub struct Interceptor {
+    label: Vec<u8>,
```
Comment (tiziano88) [inline oak_client/src/lib.rs:43]: Why not keep this as type Label?

File: oak_client/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
```
Comment (tiziano88) [inline oak_client/src/lib.rs:44]: Add a TODO somewhere to deal with authentication too?

File: oak_client/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        ClientTlsConfig::new().ca_certificate(Certificate::from_pem(root_tls_certificate.to_vec()));
+    Channel::builder(uri.clone())
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Intercepts gRPC messages and adds `label` as a gRPC metadata.
```
Comment (tiziano88) [inline oak_client/src/lib.rs:41]: Also mention that it adds the same label to each request.

**Review by tiziano88 — APPROVED** [review]

File: oak_client/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+impl Into<tonic::Interceptor> for Interceptor {
+    fn into(self) -> tonic::Interceptor {
+        let interceptor = move |mut request: Request<()>| {
+            request.metadata_mut().insert_bin(
+                oak_abi::OAK_LABEL_GRPC_METADATA_KEY,
+                MetadataValue::from_bytes(self.label.as_ref()),
+            );
+            Ok(request)
```
Comment (tiziano88) [inline oak_client/src/lib.rs:65]:
```suggestion
            Ok(())
```

File: oak_client/src/lib.rs (hunk 2 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:43]: Because encoding the label returns an error. And `Into<Interceptor>` couldn't return an error.

Comment (tiziano88) [reply oak_client/src/lib.rs:43]: That seems worth a comment in the file, right?

File: oak_client/src/lib.rs (hunk 4 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:41]: Done

File: oak_client/src/lib.rs (hunk 2 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:43]: Done

File: oak_client/src/lib.rs (hunk 3 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:44]: Done

File: oak_abi/src/label/mod.rs (hunk 1 above)
Comment (ipetr0v) [reply oak_abi/src/label/mod.rs:75]: I thought since we already have `public_untrusted` method, we just could add convenience functions for simple labels.

Comment (tiziano88) [reply oak_abi/src/label/mod.rs:75]: `public_untrusted` is like `default` or `empty`, that's why it is defined on `Label`. But I think individual label constructors can just be regular functions outside of it.

Comment (ipetr0v) [reply oak_abi/src/label/mod.rs]: Done, added:
```rust
/// Convenience function for creating labels with a single confidentiality tag.
pub fn confidentiality_label(tag: Tag) -> crate::proto::oak::label::Label {
    Label {
        confidentiality_tags: vec![tag],
        integrity_tags: vec![],
    }
}
```

**Review by rbehjati — APPROVED** [review]
LGTM :)

**Pull request merged**